### PR TITLE
Add unit tests to the client and fix related bugs

### PR DESCRIPTION
**Pull request by ghost**
This PR adds unit tests for the types and the functions in the API Client. 
When testing, several bugs were identified, leading to the implementation of the following modifications: 
* Add a missing optional keyword to the `UpdateGridpoolOrderRequest` protobuf definition
* Make the `GridpoolOrderFilter`, `GridpoolTradeFilter` and `PublicTradeFilter` types hashable, as they are used as keys in combination with the `gridpool_id` in the streaming functions
* Refactor some protofub timestamps that were falsly returning None values in the client
* Add error handling in the client for the gRPC errors that could be raised by the service

### CODE REVIEW

File: proto/frequenz/api/electricity_trading/v1/electricity_trading.proto
Comment (matthias-wende-frequenz) [inline proto/frequenz/api/electricity_trading/v1/electricity_trading.proto]: For singular messages setting or not setting optional makes no difference. They are having explicit presence semantics in both cases. 
AFAIR we agreed on removing the optional keyword everywhere where it's not needed and only document it's optional nature.
So instead of adding optional here we should remove it in line 260.

File: pytests/test_types.py
Comment (matthias-wende-frequenz) [inline pytests/test_types.py]: You are converting back and force using your own implementation under the assumption that this is correct implementation.

Imagine I would change the `to_pb` and `from_pb` methods with trivial identity functions. Then this test would pass although the functions would not fulfill their purpose. In that sense this unit test defeats it's purpose.

As a simple workaround you could test if the type of the converted type has changed to the expected type using the `isinstance` method.

Same holds true for all the following tests.

File: py/frequenz/client/electricity_trading/_types.py
Comment (matthias-wende-frequenz) [inline py/frequenz/client/electricity_trading/_types.py:1060]: Is there a reason why you've extracted this into two separate statements? Same question for the following cases.

File: py/frequenz/client/electricity_trading/_types.py
Comment (matthias-wende-frequenz) [inline py/frequenz/client/electricity_trading/_types.py:1345]: Can you please specify in the commit comment, what type of Bugs is fixed by that change?

**Review by matthias-wende-frequenz — CHANGES_REQUESTED** [review]
First round of review done. I'll review the second part tomorrow.

File: pytests/test_client.py
Comment (matthias-wende-frequenz) [inline pytests/test_client.py]: I wonder why mypy doesn't complain. You should ideally use snake case for function names.

File: pytests/test_client.py
Comment (matthias-wende-frequenz) [inline pytests/test_client.py]: same here

File: py/frequenz/client/electricity_trading/_client.py
Comment (matthias-wende-frequenz) [inline py/frequenz/client/electricity_trading/_client.py:135]: You are only logging the error. You should raise a new (or the same) error instead, such that the downstream user of the client can handle the error. Same applies for the following functions.

File: proto/frequenz/api/electricity_trading/v1/electricity_trading.proto
Comment (ghost) [reply proto/frequenz/api/electricity_trading/v1/electricity_trading.proto]: No, we had to remove the optional keywords for the types that do not have a 'sane default'. `OrderExecutionOption` should still have an optional keyword because its 'sane default' value is `UNSPECIFIED`. [This](https://github.com/frequenz-floss/frequenz-api-electricity-trading/issues/19) issue has more information about it

File: py/frequenz/client/electricity_trading/_types.py
Comment (ghost) [reply py/frequenz/client/electricity_trading/_types.py:1060]: Yes this was to be done in 2 steps else it returns none

File: proto/frequenz/api/electricity_trading/v1/electricity_trading.proto
Comment (matthias-wende-frequenz) [reply proto/frequenz/api/electricity_trading/v1/electricity_trading.proto]: You are right. I misread and thought this is a message rather then an enum. Sorry for the fuss ;).

File: pytests/test_client.py
Comment (ghost) [reply pytests/test_client.py]: (just for the record): Because this is a function inheriting from `unittest.TestCase`, but this won't matter anyways because I'm changing to `pytest` now

**Comment by ghost** [comment]
**NOTE**: The last commit ("Remove _lists suffixes in client") will be removed once the changes in #59 are merged.

File: pytests/test_client.py
Comment (idlir-shkurti-frequenz) [inline pytests/test_client.py]: This is used a couple of times in the code. Maybe its better to define as a constant somewhere at the start of the script for readability purposes.

File: pytests/test_client.py
Comment (idlir-shkurti-frequenz) [inline pytests/test_client.py:274]: Maybe a good idea to also add some tests when we expect the client to return an error or exception.

File: pytests/test_types.py
Comment (idlir-shkurti-frequenz) [inline pytests/test_types.py]: This might be a longer and maybe unnecessary change but it is something to keep in mind for the future. There are a lot of conversions to pb in this script which could be abstracted in the future into helper functions to reduce duplications and make the code a bit more readable. So you could have a function called `assert_conversion_to_pb` which takes in the original type and has an input which is the function which converts the original to the protobuf representation and asserts it (the assert function can also be a callable input of the `assert_conversion_to_pb` function.

Like I said, this is maybe not a priority for now if we want to move fast now so I will leave it up to you if you want to implement it. But maybe something to keep in mind for the future either way.

File: pytests/test_types.py
Comment (idlir-shkurti-frequenz) [inline pytests/test_types.py:534]: Maybe the same point as in `test_client.py` could be useful here. I would add some tests where we expect the types to not match. This might be a bit tougher for the types than in the client case though so not sure if it makes a lot of sense here.
For example there was a check in place [here](https://github.com/frequenz-floss/frequenz-api-electricity-trading/blob/a66061ec3a702becdafb1f1d18a22d5f20fa075a/py/frequenz/client/electricity_trading/_types.py#L959) if the order has a `valid_until` type which could be tested here also.

File: pytests/test_types.py
Comment (idlir-shkurti-frequenz) [reply pytests/test_types.py]: A little example of what I mean with the help of chatgpt:

```python
def assert_conversion_to_pb(original: Any, convert_to_pb: Callable[[], T], expected_pb: T, assert_func: Callable[[T, T], None]) -> None:
    """
    Generic utility function to test conversion from a custom type to protobuf.

    :param original: The original instance of the custom type.
    :param convert_to_pb: Function that converts `original` to its protobuf representation.
    :param expected_pb: The expected protobuf instance.
    :param assert_func: Function to assert equality between two protobuf instances.
    """
    converted_pb = convert_to_pb()
    assert_func(converted_pb, expected_pb)
```

Then we would need a specific assert function for comparing pb messages:

```python
def assert_protobuf_equal(actual_pb: Any, expected_pb: Any) -> None:
    """
    Asserts that two protobuf instances are equal, using DeepDiff for detailed comparison.

    :param actual_pb: The actual protobuf instance.
    :param expected_pb: The expected protobuf instance.
    """
    diff = DeepDiff(actual_pb, expected_pb, ignore_order=True)
    assert not diff, f"Differences found in protobuf comparison: {diff}"
```

and then we could refactor the test cases like this:

```python
def test_price_to_pb_refactored() -> None:
    """Refactored test for the client price type conversions to protobuf using DRY principle."""
    assert_conversion_to_pb(
        original=Price(amount=Decimal("100"), currency=Currency.USD),
        convert_to_pb=lambda: Price(amount=Decimal("100"), currency=Currency.USD).to_pb(),
        expected_pb=price_pb2.Price(
            amount=decimal_pb2.Decimal(value="100"),
            currency=price_pb2.Price.Currency.CURRENCY_USD,
        ),
        assert_func=assert_protobuf_equal
    )
```

File: pytests/test_types.py
Comment (ghost) [reply pytests/test_types.py:534]: (fyi I added some tests for no/ invalid timezones in`OrderDetail`)

> For example there was a check in place [here](https://github.com/frequenz-floss/frequenz-api-electricity-trading/blob/a66061ec3a702becdafb1f1d18a22d5f20fa075a/py/frequenz/client/electricity_trading/_types.py#L959) if the order has a valid_until type which could be tested here also.

Here I check that the attributes converted are the right ones (with `deepDiff`), and I make sure that no other instance var is converted (by checking the number of attributes in the converted type matches the one in the original type). You mean I should do multiple of these tests for different fields to update ?

File: pytests/test_client.py
Comment (ghost) [reply pytests/test_client.py:274]: Just for the record: I don't think it makes that much sense here since these functions are wrappers only and don't raise errors themselves. Given that I am mocking the underlying function testing an exception being raised would mean I need to mock the response of the graph function to return an error, and then checking that this error was indeed returned.

File: pytests/test_types.py
Comment (idlir-shkurti-frequenz) [reply pytests/test_types.py:534]: > You mean I should do multiple of these tests for different fields to update ?

Yeah ideally but I think we can always increase the test coverage so maybe this is enough for now if we want to focus on deploying soon.

**Review by idlir-shkurti-frequenz — DISMISSED** [review]
LGTM!

File: pytests/test_types.py
Comment (matthias-wende-frequenz) [inline pytests/test_types.py:365]: I wonder if we also should check if the time has been shifted according to the tz change.

File: RELEASE_NOTES.md
Comment (matthias-wende-frequenz) [inline RELEASE_NOTES.md]: Seems that this has been done in a different PR. Correct?

File: RELEASE_NOTES.md
Comment (matthias-wende-frequenz) [inline RELEASE_NOTES.md]: And other types too ;)

**Review by matthias-wende-frequenz — COMMENTED** [review]
Some more smallish comments. Otherwise this looks fine.

File: RELEASE_NOTES.md
Comment (camille-bouvy-frequenz) [reply RELEASE_NOTES.md:28]: It made the tests fail that's why it's here

File: pytests/test_types.py
Comment (ghost) [reply pytests/test_types.py:365]: Can do but imo I don't think it really makes sense given the way the conversion is done

Comment (matthias-wende-frequenz) [reply pytests/test_types.py:365]: How the conversion is done is implementation specific. The test doesn't know anything about that and the implementation might change in the future.

File: RELEASE_NOTES.md
Comment (ghost) [reply RELEASE_NOTES.md]: True my bad this was subsequently done by you in #61 actually so I can remove it here

File: pytests/test_types.py
Comment (ghost) [reply pytests/test_types.py:365]: Yeah fair point, I added a line now to check

File: RELEASE_NOTES.md
Comment (matthias-wende-frequenz) [reply RELEASE_NOTES.md]: Oh great stuff :D

**Review by matthias-wende-frequenz — APPROVED** [review]

**Pull request merged**
